### PR TITLE
Feature/remove 3rd header datatypes and using the custom graph response in cpp

### PR DESCRIPTION
**Pull request by Shylock-Hg**

### CODE REVIEW

**Review by laura-ding — APPROVED** [review]

File: src/common/datatypes/Edge.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+// Inject a customized hash function
+std::size_t hash<nebula::Edge>::operator()(const nebula::Edge& h) const noexcept {
+    size_t hv = folly::hash::fnv64(h.src);
+    hv = folly::hash::fnv64(h.dst, hv);
+    hv = folly::hash::fnv64_buf(reinterpret_cast<const void*>(&h.type),
+                                sizeof(h.type),
+                                hv);
+    return folly::hash::fnv64_buf(reinterpret_cast<const void*>(&h.ranking),
+                                    sizeof(h.ranking),
```
Comment (darionyaphet) [inline src/common/datatypes/Edge.cpp:46]: alignment ?

File: src/common/datatypes/DataSet.h [hunk 2]
```diff
@@ -7,7 +7,9 @@
 #ifndef COMMON_DATATYPES_DATASET_H_
 #define COMMON_DATATYPES_DATASET_H_
 
-#include "common/base/Base.h"
+#include <iostream>
```
Comment (darionyaphet) [inline src/common/datatypes/DataSet.h:10]: reduce the useless include ?

File: src/common/datatypes/Date.cpp [hunk 3]
```diff
@@ -126,3 +131,56 @@ std::string DateTime::toString() const {
...
+                                sizeof(h.hour),
+                                hv);
+    hv = folly::hash::fnv64_buf(reinterpret_cast<const void*>(&h.minute),
+                                sizeof(h.minute),
+                                hv);
+    hv = folly::hash::fnv64_buf(reinterpret_cast<const void*>(&h.sec),
+                                sizeof(h.sec),
+                                hv);
+    return folly::hash::fnv64_buf(reinterpret_cast<const void*>(&h.microsec),
+                                sizeof(h.microsec),
```
Comment (darionyaphet) [inline src/common/datatypes/Date.cpp:182]: alignment

Comment (Shylock-Hg) [reply src/common/datatypes/Date.cpp]: Thanks.Fixed.

File: src/common/datatypes/Edge.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/datatypes/Edge.cpp]: Thanks.Fixed.

File: src/common/datatypes/DataSet.h (hunk 2 above)
Comment (Shylock-Hg) [reply src/common/datatypes/DataSet.h:10]: Used in line 133

File: src/common/clients/graph/GraphClient.cpp [hunk 4]
```diff
@@ -74,21 +74,21 @@ void GraphClient::disconnect() {
 }
 
 
-cpp2::ErrorCode GraphClient::execute(folly::StringPiece stmt,
-                                     cpp2::ExecutionResponse& resp) {
+nebula::ErrorCode GraphClient::execute(folly::StringPiece stmt,
+                                     nebula::ExecutionResponse& resp) {
```
Comment (laura-ding) [inline src/common/clients/graph/GraphClient.cpp:78]: alignment

File: src/common/graph/tests/ResponseEncodeDecodeTest.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+#include "common/graph/Response.h"
+
+namespace nebula {
+
+using serializer = apache::thrift::CompactSerializer;
+
+TEST(ResponseEncodDecodeTest, Basic) {
+    // auth response
+    {
+        AuthResponse resp {};
```
Comment (laura-ding) [inline src/common/graph/tests/ResponseEncodeDecodeTest.cpp:19]: set useful value

File: src/common/graph/tests/ResponseEncodeDecodeTest.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        buf.reserve(128);
+        serializer::serialize(resp, &buf);
+        AuthResponse copy;
+        std::size_t s = serializer::deserialize(buf, copy);
+        ASSERT_EQ(s, buf.size());
+        EXPECT_EQ(resp, copy);
+    }
+    // response
+    {
+        ExecutionResponse resp {};
```
Comment (laura-ding) [inline src/common/graph/tests/ResponseEncodeDecodeTest.cpp:30]: ditto

File: src/common/clients/graph/GraphClient.cpp (hunk 4 above)
Comment (Shylock-Hg) [reply src/common/clients/graph/GraphClient.cpp]: Thanks.Fixed.

File: src/common/graph/tests/ResponseEncodeDecodeTest.cpp (hunk 5 above)
Comment (Shylock-Hg) [reply src/common/graph/tests/ResponseEncodeDecodeTest.cpp]: Ok, fixed.

File: src/common/graph/Response.h [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        description = nullptr;
+        profiles = nullptr;
+        branch_info = nullptr;
+        dependencies = nullptr;
+    }
+
+    bool operator==(const PlanNodeDescription &rhs) const;
+
+    std::string                                   name;
+    int64_t                                       id;
```
Comment (laura-ding) [inline src/common/graph/Response.h:147]: init

File: src/common/graph/Response.h [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        if (!checkPointer(error_msg.get(), rhs.error_msg.get())) {
+            return false;
+        }
+        if (!checkPointer(plan_desc.get(), rhs.plan_desc.get())) {
+            return false;
+        }
+        return checkPointer(comment.get(), rhs.comment.get());
+    }
+
+    ErrorCode error_code{ErrorCode::SUCCEEDED};
```
Comment (laura-ding) [inline src/common/graph/Response.h:211]: var format

File: src/common/graph/Response.h [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        is_do_branch = false;
+        condition_node_id = -1;
+    }
+
+    bool operator==(const PlanNodeBranchInfo &rhs) const {
+        return is_do_branch == rhs.is_do_branch && condition_node_id == rhs.condition_node_id;
+    }
+
+    // True if loop body or then branch of select
+    bool  is_do_branch;
```
Comment (laura-ding) [inline src/common/graph/Response.h:114]: init

File: src/common/graph/Response.h [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    // select/loop node id
+    int64_t  condition_node_id;
```
Comment (laura-ding) [inline src/common/graph/Response.h:116]: ditto

File: src/common/graph/Response.h (hunk 8 above)
Comment (Shylock-Hg) [reply src/common/graph/Response.h:211]: This name should keep same with thrift.

File: src/common/graph/Response.h (hunk 7 above)
Comment (Shylock-Hg) [reply src/common/graph/Response.h]: Good catch. Fixed.

File: src/common/graph/Response.h (hunk 8 above)
Comment (Shylock-Hg) [reply src/common/graph/Response.h]: Fixed.

File: src/common/clients/graph/GraphClient.cpp [hunk 11]
```diff
@@ -28,8 +28,8 @@ GraphClient::~GraphClient() {
 }
 
 
-cpp2::ErrorCode GraphClient::connect(const std::string& username,
-                                     const std::string& password) {
+nebula::ErrorCode GraphClient::connect(const std::string& username,
```
Comment (laura-ding) [inline src/common/clients/graph/GraphClient.cpp:31]: I think you can delete the GraphClient.cpp and GraphClient.h

Comment (Shylock-Hg) [reply src/common/clients/graph/GraphClient.cpp:31]: Yes, when cpp client ready.

**Review by laura-ding — DISMISSED** [review]

File: src/common/datatypes/Value.h [hunk 12]
```diff
@@ -307,11 +305,12 @@ struct Value {
...
 
-    StatusOr<bool> toBool();
+    // TODO(shylock) try optional when c++17
+    std::unique_ptr<bool> toBool();
 
-    StatusOr<double> toFloat();
+    std::unique_ptr<double> toFloat();
 
-    StatusOr<int64_t> toInt();
+    std::unique_ptr<int64_t> toInt();
```
Comment (dutor) [inline src/common/datatypes/Value.h:313]: Use `std::pair<T, bool>`.

File: src/common/graph/AuthResponseOps.h [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+namespace apache {
+namespace thrift {
+
+/**************************************
+ *
+ * Ops for class AuthResponse
+ *
+ *************************************/
+namespace detail {
```
Comment (dutor) [inline src/common/graph/AuthResponseOps.h:27]: Why is this needed?

**Review by dutor — COMMENTED** [review]
Excellent job

**Review by dutor — APPROVED** [review]
LGTM.

**Review by laura-ding — APPROVED** [review]
